### PR TITLE
Support for correcting invoices

### PR DESCRIPTION
**Pull request by samlown**
* Preliminary support for correcting invoices.
* Will automatically determine what is possible based on the Regime configuration.
* Automatically inverts quantities of Credit Notes are not supported locally.
* Uses the `PrecedingDefinition` structure to standardise expectations (as opposed to local methods).
* TODO: validation at a regime level.
* TODO: determine more examples.

### CODE REVIEW

**Review by cavalle — APPROVED** [review]
I think this is a very cool functionality!! I'm just crossing my fingers 🤞 that the model is simple and universal enough so that we can scale this nicely to many more countries. I have no reasons to believe it won't at this point, but there's a ton I ignore about how invoices are corrected in most countries!

I just left a few comments in the code. Beside those, I was wondering if we should also add support for PT where we're already able to issue invoice credit notes. An implementation like the Colombian one would work. [Here](https://github.com/invopop/gobl/blob/main/regimes/pt/pt.go#L52) is the stamp,

**Comment by samlown** [comment]
> I'm just crossing my fingers 🤞 that the model is simple and universal enough so that we can scale this nicely to many more countries.

Agreed. My initial implementation was way more flexible, and complex, but then it became clear that each of the countries had a pretty similar set of scenarios and it could be standardised. Let's see how it goes!

Great suggestions, I'll go through commenting and implementing them now.

File: bill/invoice_correct.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		Series:           inv.Series,
+		Code:             inv.Code,
+		IssueDate:        &inv.IssueDate,
+		Reason:           o.Reason,
+		Corrections:      o.Corrections,
+		CorrectionMethod: o.CorrectionMethod,
+	}
+
+	// Take the regime def to figure out what needs to be copied
+	if o.Refund {
```
[the remark this answers is not in the document]
Comment (samlown) [reply bill/invoice_correct.go:104]: "Refund" felt closer to the mental action the user will be performing: "I need to refund this invoice", but revoke is probably more accurate, as you say, it doesn't imply payment. I'd also considered `Credit`, which is more closely aligned with a Credit Note. I think internally `Credit` might be better, as `Revoke` might give the impression that a secondary action is happening. WDYT?

File: bill/invoice_correct.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+			// regular credit note
+			i2.Type = InvoiceTypeCreditNote
+		} else if InvoiceTypeCorrective.In(r.Preceding.Types...) {
+			// corrective invoice with negative values
+			i2.Type = InvoiceTypeCorrective
+			i2.Invert()
+		} else {
+			return nil, errors.New("refund not supported by regime")
+		}
+	} else if o.Append {
```
[the remark this answers is not in the document]
Comment (samlown) [reply bill/invoice_correct.go:115]: Well spotted! There is a bug in the code, I forgot to add the call to the new `Empty()` method that clears out all the lines, charges, discounts, and outlays. That's fixed now.

Regarding mutual exclusivity, the Append and Refund options are like this as I was thinking of the scenario of flags on the command line:

```
gobl correct --refund in.json
```

A quick flag is way easier to use than a key-value combo. Would you be happy if I at least add an error check :-)

File: bill/invoice_correct.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		if InvoiceTypeDebitNote.In(r.Preceding.Types...) {
+			// regular debit note
+			i2.Type = InvoiceTypeDebitNote
+		} else {
+			return nil, errors.New("append not supported by regime")
+		}
+	} else {
+		if InvoiceTypeCorrective.In(r.Preceding.Types...) {
+			i2.Type = InvoiceTypeCorrective
+		} else {
```
[the remark this answers is not in the document]
Comment (samlown) [reply bill/invoice_correct.go:125]: So for this case, without the "Refund" or "Append" flags, the idea is that you're issuing a new invoice that simply replaces the old. In Spain, this is allowed, but does not appear to be the case in other countries.

My suggestion to users wanting to issue a partial credit-note, would be tu use the refund flag and simply remove the rows they don't need to refund... make sense?

File: regimes/co/co.go [hunk 4]
```diff
@@ -51,6 +51,13 @@ func New() *tax.Regime {
 		Calculator: Calculate,
 		Zones:      zones, // see zones.go
 		Preceding: &tax.PrecedingDefinitions{ // see preceding.go
+			Types: []cbc.Key{
+				bill.InvoiceTypeCreditNote,
+			},
+			Stamps: []cbc.Key{
+				StampProviderDIANCUDE,
+				StampProviderDIANQR,
```
[the remark this answers is not in the document]
Comment (samlown) [reply regimes/co/co.go:59]: Cool. Will remove 👍

**Comment by samlown** [comment]
@cavalle @lucanioi this turned into a mega-PR, sorry. Key take aways: we have two flags: `Credit` and `Debit`, which are mutually exclusive (an error will be raised if both used), but have a nice API:

```go
i2 := i.Correct(bill.Credit)
```

Let me know if you feel strongly about this and I'll refactor.

I also got a bit carried away refactoring some of the regimes to support the preceding definitions and ended up adding tax code validation for France and Great Britain. I came across this library which will help to copy and paste more local tax code validation: https://github.com/ltns35/go-vat (which is itself based off a JS version).

File: bill/invoice_correct.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cavalle) [reply bill/invoice_correct.go]: I like "Credit" too!

File: bill/invoice_correct.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cavalle) [reply bill/invoice_correct.go]: The error check would makes sense to me :+1:

File: bill/invoice_correct.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cavalle) [reply bill/invoice_correct.go:131]: Yes, agreed, makes sense. "Credit" (full automatically, partial if edited manually), "Append" and "Replace" (implicit, by default) would be the three behaviours.

**Comment by cavalle** [comment]
> Key take aways: we have two flags: Credit and Debit, which are mutually exclusive (an error will be raised if both used), but have a nice API:

Credit and Debit sound great to me! I think that's the API we're looking for. I say let's go for it!

**Pull request merged**